crates/bitwarden-vault/src/cipher/repository.rs

```diff
@@ -0,0 +1,212 @@
+use std::sync::Arc;
+
+use bitwarden_core::{require, MissingFieldError};
+use bitwarden_db::{named_params, params, Database, DatabaseError, DatabaseTrait, RowError};
+use thiserror::Error;
+use tokio::sync::Mutex;
+use uuid::Uuid;
+
+use super::Cipher;
+
+/// A row in the ciphers table.
+struct CipherRow {
+    #[allow(dead_code)]
+    id: Uuid,
+    value: String,
+}
+
+#[derive(Debug, Error)]
+pub enum CipherRepositoryError {
+    #[error(transparent)]
+    MissingFieldError(#[from] MissingFieldError),
+    #[error(transparent)]
+    DatabaseError(#[from] DatabaseError),
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::Error),
+}
+
+pub struct CipherRepository {
+    db: Arc<Mutex<Database>>,
+}
+
+impl CipherRepository {
+    pub fn new(db: Arc<Mutex<Database>>) -> Self {
+        Self { db: db.clone() }
+    }
+
+    pub async fn save(&self, cipher: &Cipher) -> Result<(), CipherRepositoryError> {
+        let id = require!(cipher.id);
+        let serialized = serde_json::to_string(cipher)?;
+
+        let guard = self.db.lock().await;
+
+        Ok(guard
+            .execute(
+                "
+                    INSERT INTO ciphers (id, value)
+                    VALUES (?1, ?2)
+                    ON CONFLICT(id) DO UPDATE SET
+                    value = ?2
+                ",
+                params![&id, &serialized],
+            )
+            .await
+            .map(|_| ())?)
+    }
+
+    pub async fn replace_all(&self, ciphers: &[Cipher]) -> Result<(), CipherRepositoryError> {
+        let guard = self.db.lock().await;
+
+        guard.execute("DELETE FROM ciphers", []).await?;
```

Comment:
Would be nice to be able to run the `DELETE FROM` and `INSERT INTO` in the same transaction, to be able to rollback the whole thing in cases there's an error.